doc/man/pvfs2-change-fsid.1
.\" Process this file with
.\" groff -man -Tascii foo.1
.\" 
.TH "PVFS2-CHANGE-FSID" "1" "SEPTEMBER 2011" "PVFS2" "PVFS2 MANUALS"
.SH "NAME"
pvfs2\-change\-fsid \- Utility for updating fsid in PVFS2 collections
.SH "SYNOPSIS"
.B pvfs2\-change\-fsid  [\-\-verbose] [\-\-oldfsid fsid]
.I [\-\-newfsid fsid] [\-\-fsname fsname]
.I [\-\-view]
.I \-\-dbpath=</path/to/collections.db> \-\-fsconf=</path/to/pvfs2\-fs.conf>
.I \-\-storage=</path/to/pvfs2\-storage\-space>
.SH "DESCRIPTION"
Changes the fsid of the file system in the database given as the \-\-dbpath. If
the old fsid is not provided then it will be looked up in the conf file.
Likewise, if a new fsid is not specified then the new fsid will be generated in
the same manner as pvfs2\-genconfig. Attempting to set the new fsid to the 
current fsid will cause the program to exit silently without making any changes.
Changes to the fsid will automatically be updated in the conf file.
.SH "OPTIONS"
.IP \-\-dbpath
The path to the collections.db that contains the file system ID that you want to change.
.IP \-\-fsconf
The path to the pvfs2 configuration file for the file system being modified.
.IP \-\-storage
Path to the local stroage space for the file system being modified.
.IP \-\-oldfsid
The current file system ID. If noy supplied then it will be looked up from the configuration
files.
.IP \-\-newfsid
The ID that you would like to change to. If not supplied then it will be generated as
in genconfig.
.IP \-\-view
Don't make the changes to the file system, instead just show what those changes would be.
.IP \-\-verbose
Print verbose messages.
.IP \-\-help
Show a basic help listing.
.SH "ENVIRONMENT"
.IP PVFS2TAB_FILE
If set, the full pathname for an alternate 
.I R pvfs2tab
file

.SH "BUGS"
Please report all bugs to the pvfs2 bugzilla (http://bugzilla.mcs.anl.gov/pvfs/)
.SH "AUTHOR"
The PVFS2 development team (pvfs2\-developers@beowulf\-underground.org)


ChangeLog
-----------------------
OrangeFS Release ChangeLog
-----------------------
orangefs-2.8.5

Text after "CVS: " can be prepended with "http://www.pvfs.org/fisheye/changelog/~br=Orange-Branch/PVFS/?cs=Orange-Branch:" to view related change sets in FishEye.

* Enhancement: change to kernel source path to support SuSE style kernel sources
        CVS: mtmoore:20110831005433

* Enhancement: add prompt after X removals during a pvfs2-fsck when -y specified
        CVS: mtmoore:20110824144127

* Enhancement: Add sanity checks to extended attribute operations 
        CVS: mtmoore:20110811180653

* Enhancement addition of counters and application to feed SNMP counters
        CVS: walt:20110628184842
        CVS: walt:20110628161840
        CVS: walt:20110617152339

* Enhancement: addition of scripts used for OrangeFS CI Jenkins testing
        CVS: mtmoore:20110623183144

* Enhancement: add backtrace handler to SIGABRT
        CVS: mtmoore:20110607235041

* Enhancement: add sanity checks during trove cleanup/initialization
        CVS: mtmoore:20110607232817

* Enhancement: allow server log to be re-opened on SIGHUP to allow log rotation
        CVS: mtmoore:20110607023901        
        CVS: mtmoore:20110530143543

* Enhancement: allow debug mask to be reset during SIGHUP
        CVS: bligon:20110504200411

* Enhancement: add Berkeley DB configure check for version 4.8.30
        CVS: mtmoore:20110405050554
	Additional Information: https://www.orangefs.org/trac/orangefs/wiki/DBVersionRequired

* Enhancement add configure checks for binaries required during build
        CVS: mtmoore:20110404165443

* Enhancement: kernel changes to support newer kernels (up to 2.6.40 tested and 3.X compiles, not tested)
        CVS: mtmoore:20110831005433
        CVS: mtmoore:20110823184109
        CVS: mtmoore:20110822232017
        CVS: mtmoore:20110412175323
        CVS: mtmoore:20110525205504
        CVS: mtmoore:20110405142136
        CVS: mtmoore:20110405141506

* Bug Fix: man pages are now gzip'd during installation  
        CVS: mtmoore:20110904010919

* Bug Fix: berkeley db gives back unaligned pages (in custom compare/multi-key cases), use local variables to fix ARM failures
        CVS: mtmoore:20110902173424

* Bug Fix: pointer dereference size issue in dbpf-mgmt on non-x86 platforms, Neal Ferguson
        CVS: mtmoore:20110823171834

* Bug Fix: file corruption using direct backend when writes continuously appending to single file
        CVS: elaine:20110819120934
        CVS: mtmoore:20110629130442

* Bug Fix: segfault in pvfs2-xattr due to mirror handle string comparison, Nick Mills
        CVS: mtmoore:20110809180325

* Bug Fix: add backtrace functionality back in for newer distributions
        CVS: mtmoore:20110801124552
        CVS: mtmoore:20110607215819

* Bug Fix: server-side memory leak related to operation completion 
        CVS: bligon:20110708194724

* Bug Fix: slab cache fix, Bart Taylor
        CVS: bligon:20110613200154

* Bug Fix: duplicate handle fix in iterate-handles, Phil Carns
        CVS: bligon:20110610202309

* Bug Fix: correctly parse pvfs2-server alias argument
        CVS: mtmoore:20110608010721

* Bug Fix: mx_address lookup failure with multiple protocols
        CVS: bligon:20110530173949

* Bug Fix: admin apps segfault with IB
        CVS: bligon:20110512215357

* Bug Fix: allow pvfs2-server to re-read configuratio file regardless of absolute or relative path
        CVS: mtmoore:20110506194814

* Bug Fix: root squash functionality, Bart Taylor
        CVS: bligon:20110506173620

* Bug Fix: various kernel related stability and error-handling fixes
        CVS: mtmoore:20110729151843
        CVS: mtmoore:20110729143317
        CVS: mtmoore:20110530154853
        CVS: mtmoore:20110412175444

* Bug Fix: pin memory passed in to the kernel to prevent swap pages and kernel panic
        CVS: mtmoore:20110725160953

-------------------------------------------------------------------------------

orangefs-2.8.4
* Change verisioning to no longer include dates, just increment the sub version for every release

orangefs-2.8.3-20110323
* Bug fix: resolve assert in dbpf_open_cache_remove() when using tree code for file remove
        CVS: Orange-Branch:bligon:20110322141529

* Bug fix: PINT_dist_encode() segfault in pvfs2-server caused by the client create state machine not sending dist information on create retries.
        CVS: Orange-Branch:mtmoore:20110322014656 
        CVS: Orange-Branch:mtmoore:20110322031703
        CVS: Orange-Branch:mtmoore:20110322040628
        CVS: Orange-Branch:mtmoore:20110322114637
        CVS: Orange-Branch:mtmoore:20110322121208 

* Bug fix: fix incorrect attribute output using sys attrs
        CVS: Orange-Branch:mtmoore:20110301023749

* Bug fix: fix segfault in pvfs2-server if client provides too large of buffer in the list-eattr state machine. Includes additional sanity checks on extended attribute operations
        CVS: Orange-Branch:mtmoore:20110225222047

* Bug fix: merge dir.c fix from main branch
        CVS: Orange-Branch:bligon:20110126215228 

* Bug fix: resolve data corruption caused by ki_pos not being updated during synchronous I/O through asynchronous path (relevant to newer kernels).
        CVS: Orange-Branch:mtmoore:20110123143300

* Bug fix: fix resolving symbolic links from pvfs2-* utilities
        CVS: Orange-Branch:mtmoore:20110114215452 

* Bug fix: Fix ncache entry update when moving files by David Metheny
        CVS: Orange-Branch:mtmoore:20110111154332 

* Bug fix: PVFS_hint_add fix by Julian Kunkel and refinement of patch
        CVS: Orange-Branch:mtmoore:20110105051448 
	CVS: Orange-Branch:bligon:20110224203131

* Bug fix: change where unlocks occur w.r.t __PVFS_JOB_THREADED__ to prevent deadlock
        CVS: Orange-Branch:mtmoore:20101221163004

* Bug fix: change incorrect use of union members in tree-communicate
        CVS: Orange-Branch:mtmoore:20101214182035 


src/apps/admin/pvfs2-change-fsid.c
/*
 * (C) 2001 Clemson University and The University of Chicago
 *
 * See COPYING in top-level directory.
 */

/** \file
 * Update utility for updating fsid in PVFS2 collections
 */

#include <unistd.h>
#include <string.h>
#include <stdio.h>
#include <getopt.h>
#include <limits.h>
#include <string.h>
#include <stdlib.h>
#include <errno.h>
#include <sys/types.h>
#include <dirent.h>
#include <limits.h>
#include <inttypes.h>
#include <time.h>

#include <db.h>

#include "pvfs2-config.h"
#include "pvfs2.h"
#include "pvfs2-internal.h"
#include "trove.h"
#include "mkspace.h"
#include "pint-distribution.h"
#include "pint-dist-utils.h"


typedef struct
{
    char db_path[PATH_MAX];
    char fs_conf[PATH_MAX];
    char fs_name[PATH_MAX];
    char storage_path[PATH_MAX];
    int32_t old_fsid;
    int32_t new_fsid;
    char old_fsid_hex[9];
    char new_fsid_hex[9];
    int verbose;
    int view_only;
    int hex_dir_exists;
} options_t;


int update_fs_conf(void);
int update_fsid_in_collections_db(void);
int get_old_fsid_from_conf(void);
int move_hex_dir(void);
int process_args(int argc, char ** argv);
int setup(int argc, char ** argv);
void print_help(char * progname);

static options_t opts;

int main(int argc, char ** argv)
{
    int ret = 0;
    
    ret = setup( argc, argv);
    if(ret)
    {
        printf("Error in setup function\n");
        return -1;
    }

    if(opts.old_fsid == opts.new_fsid)
    {
        if(opts.verbose)
        {
            printf("Nothing to do. Old/New fsids are the same.\n");
        }
        return 0;
    }

    /* Fix up the pvfs2-fs.conf file */
    if(!opts.view_only)
    {
        ret = update_fs_conf();
        if(ret)
        {
            fprintf(stderr,"Error updating %s\n", opts.fs_conf);
            return -1;
        }

        if(opts.verbose)
        {
            printf("Successfully changed ID in [%s] from [%" PRId32 "] " \
                   "to [%" PRId32 "]\n",
                   opts.fs_conf,
                   opts.old_fsid,
                   opts.new_fsid);
        }
    }

    /* Move the hex dir in the storage space */
    ret = move_hex_dir();
    if(ret)
    {
        fprintf(stderr,"Error moving hex directory in storage space.\n");
        return -1;
    }

    if(opts.view_only)
    {
        printf("ID field in [%s] is dec=[%" PRId32 "] hex=[%s]\n",
               opts.fs_conf, 
               opts.old_fsid,
               opts.old_fsid_hex);
        if(opts.hex_dir_exists)
        {
            printf("Found directory [%s] in storage space.\n", 
                   opts.old_fsid_hex);
        }
        else
        {
            printf("Directory [%s] was NOT found in storage space.\n",
                   opts.old_fsid_hex);
        }
    }

    /* Fix up the collections.db file */
    ret = update_fsid_in_collections_db();
    if(ret)
    {
        fprintf(stderr,"Error updating collections.db file\n");
        return -1;
    }

    return 0;
}

int setup( int argc, char ** argv)
{
    int ret = 0;

    /* Process command line args */
    ret = process_args( argc, argv);
    if(ret)
    {
        fprintf(stderr,"Error processing arguments\n");
        return -1;
    }

   
    /* If no old_fsid provided, look it up in the conf file */
    /*   OR if view_only is specified, look it up */
    if(!opts.old_fsid || opts.view_only)
    {
        ret = get_old_fsid_from_conf();
        if(ret)
        {
            fprintf(stderr,"Error getting fsid.\n");
            return -1;
        }
    }

    /* If no new_fsid provided, generate one */
    if(!opts.new_fsid && !opts.view_only)
    {
        srand( time(NULL) );
        /* This number came from genconfig */
        opts.new_fsid = (int32_t)abs(rand() * 2147483647);
        sprintf(opts.new_fsid_hex,"%08" PRIx32, opts.new_fsid);
        if(opts.verbose)
        {
            printf("Generated new fsid of %" PRId32 "\n", opts.new_fsid);
        }
    }

    if(opts.verbose)
    {
        printf("Moving from fsid dec=[%" PRId32 "] hex=[%s] " \
               "to fsid dec=[%" PRId32 "] hex=[%s]\n",
               opts.old_fsid,
               opts.old_fsid_hex,
               opts.new_fsid,
               opts.new_fsid_hex);
    }
     
    return 0;
}

int update_fs_conf(void)
{
    FILE * fptr = NULL;
    char command[PATH_MAX];
    char output[PATH_MAX];
    char file[512][512];
    struct stat buf;
    int ret = 0, i = 0, j = 0;
    
    /* See if fs_conf file exists */
    ret = stat(opts.fs_conf, &buf);
    if(ret)
    {
        fprintf(stderr,
                "Error checking for file's existance. [%s]",
                opts.fs_conf);
        return -1;
    }

    memset(command,0,sizeof(command));
    memset(output,0,sizeof(output));

    /* See if old_fsid is in the file provided */
    sprintf(command, 
            "grep ID %s | grep %" PRId32, 
            opts.fs_conf, 
            opts.old_fsid);
    fptr = popen(command, "r");
    if(fptr == NULL)
    {
        fprintf(stderr,"Error opening pipe. errno=%d",errno);
        exit(-1);
    }
    ret = fscanf(fptr, "%s", output);
    if(ret != 1 || !strncmp(output,"",PATH_MAX))
    {
        printf("fsid [%" PRId32 "] not found in file\n",opts.old_fsid);
        return -1;
    }
    pclose(fptr);

    memset(output,0,sizeof(output));

    /* Replace old_fsid with new_fsid and write file back out */
    sprintf(command, 
            "sed s/%" PRId32 "/%" PRId32 "/ %s", 
            opts.old_fsid, 
            opts.new_fsid, 
            opts.fs_conf);
    fptr = popen(command, "r");
    if(fptr == NULL)
    {
        fprintf(stderr,"Error opening pipe. errno=%d",errno);
        exit(-1);
    }

    i = 0;
    while( fgets(output, sizeof(output), fptr) )
    {
        strncpy(file[i],output,512);
        i++;
    }
    pclose(fptr);

    fptr = fopen(opts.fs_conf,"w");
    for(j = 0; j < i; j++)
    {
        fprintf(fptr,"%s",file[j]);
    }
    fclose(fptr);

    return 0;
}

int get_old_fsid_from_conf(void)
{
    FILE * fptr = NULL;
    int i = 0;
    char buffer[512][512];
    int32_t read_fsid = 0;
    int ret;

    memset(buffer, 0, sizeof(buffer));
    
    fptr = fopen(opts.fs_conf,"r+");
    if(fptr == NULL)
    {
        fprintf(stderr,"Error opening %s\n",opts.fs_conf);
        return -1;
    }

    /* Read in file */
    while(!feof(fptr))
    {
        ret = fscanf(fptr,"%s",buffer[i]);
        if(ret == 1)
        {
            i++;
        }
    }
    fclose(fptr);

    /* Search for ID field */
    i = 0; 
    while(strncmp(buffer[i],"ID",2))
    {
        i++;
    }
    
    /* Read in ID */
    sscanf(buffer[++i],"%" SCNd32, &read_fsid);
    if(!read_fsid) 
    {
        fprintf(stderr,"Error: fsid not found.\n");
        return -1;
    }
    
    opts.old_fsid = read_fsid;
    sprintf(opts.old_fsid_hex,"%08" PRIx32,opts.old_fsid);

    return 0;
}

int move_hex_dir(void)
{
    FILE * fptr = NULL;
    char command[PATH_MAX];
    char output[PATH_MAX];
    struct stat buf;
    char path[PATH_MAX];
    char new_path[PATH_MAX];
    int ret = 0;
    
    memset(path,0,sizeof(path));
    sprintf(path,"%s/%s", opts.storage_path, opts.old_fsid_hex);

    /* See if directory exists */
    ret = stat(path, &buf);
    if(ret)
    {
        fprintf(stderr,
                "Error checking for directory's existance. [%s]\n",
                path);
        return -1;
    }

    opts.hex_dir_exists = 1;

    if(opts.view_only)
    {
        return 0;
    }
    
    memset(command,0,sizeof(command));
    memset(output,0,sizeof(output));
    memset(new_path,0,sizeof(new_path));

    /* Move the directory */
    sprintf(new_path, "%s/%s", opts.storage_path, opts.new_fsid_hex);
    sprintf(command, "mv %s %s", path, new_path);

    fptr = popen(command, "r");
    if(fptr == NULL)
    {
        fprintf(stderr,"Error opening pipe. errno=%d",errno);
        exit(-1);
    }
    ret = fscanf(fptr, "%s", output);
    if(ret && strncmp(output,"",PATH_MAX))
    {
        printf("mv from [%s] to [%s] failed.\n", path, new_path);
        return -1;
    }
    pclose(fptr);

    if(opts.verbose)
    {
        printf("Successful dir move from [%s] to [%s]\n", path, new_path);
    }

    return 0;
}

int update_fsid_in_collections_db(void)
{
    int ret = -1;
    DB * dbp;
    DBT key, data;
    DBC * dbc_p = NULL;
    TROVE_coll_id coll_id;
   
    ret = db_create(&dbp, NULL, 0);
    if(ret != 0)
    {
        fprintf(stderr, "Error: db_create: %s.\n", db_strerror(ret));
        return(-1);
    }

    /* open collections.db */
    ret = dbp->open(dbp,
                    #ifdef HAVE_TXNID_PARAMETER_TO_DB_OPEN
                    NULL,
                    #endif
                    opts.db_path,
                    NULL,
                    DB_UNKNOWN,
                    0,
                    0);
    if(ret != 0)
    {
        fprintf(stderr,"Error:dbp->open:%s.\n", db_strerror(ret));
        return(-1);
    }

    ret = dbp->cursor(dbp, NULL, &dbc_p, 0);
    if (ret != 0)
    {
        fprintf(stderr, "Error: dbp->cursor: %s.\n", db_strerror(ret));
        dbp->close(dbp, 0);
        return(-1);
    }

    memset(&key, 0, sizeof(key));
    memset(&data, 0, sizeof(data));

    key.data = opts.fs_name;
    key.size = strlen(opts.fs_name) + 1;
    data.data = &coll_id;
    data.ulen = sizeof(coll_id);
    data.flags = DB_DBT_USERMEM;

    ret = dbp->get(dbp, NULL, &key, &data, 0);
    if (ret != 0)
    {
        fprintf(stderr, "Error: dbp->get: %s\n", db_strerror(ret));
        return -1;
    }
    
    if(opts.view_only)
    {
        printf("DB entry in [%s] for key [%s] is dec=[%" PRId32 "] " \
               "hex=[%08" PRIx32 "]\n", 
               opts.db_path, 
               opts.fs_name, 
               (int32_t)coll_id,
               (int32_t)coll_id);
        return 0;
    }

    if(opts.verbose)
    {
        printf("Retrieved key [%s] from db. Old value is [%" PRId32 "]\n", 
               (char *)key.data, 
               (int32_t)coll_id);
    }

    if(coll_id != opts.old_fsid)
    {
        fprintf(stderr, "Error: fsid retrieved does not equal old " \
                        "fsid provided! Found:[%" PRId32 "] " \
                        "Provided:[%" PRId32 "]\n",
                        (int32_t)coll_id, 
                        opts.old_fsid);
        return -1;
    }

    /* At this point the fsid (or coll_id) has been retrieved and checked */
    /*  against the user provided fsid (old) */
    /* Replace old fsid with new fsid */

    data.data = &(opts.new_fsid);
    data.ulen = sizeof(opts.new_fsid);
    data.flags = DB_DBT_USERMEM;

    ret = dbp->put(dbp, NULL, &key, &data, 0);
    if (ret != 0)
    {
        fprintf(stderr, "Error: dbp->get: %s\n", db_strerror(ret));
        return -1;
    }

    /* At this point the new fsid should be properly inserted into the db */
    /* Sanity check by retrieving again, and comparing to the new fsid */

    data.data = &coll_id;
    data.ulen = sizeof(coll_id);
    data.flags = DB_DBT_USERMEM;

    ret = dbp->get(dbp, NULL, &key, &data, 0);
    if (ret != 0)
    {
        fprintf(stderr, "Error: dbp->get: %s\n", db_strerror(ret));
        return -1;
    }

    if(coll_id != opts.new_fsid)
    {
        fprintf(stderr, "Error: fsid retrieved after the replace finished " \
                        "does not equal old fsid provided! "\
                        "Found:[%" PRId32 "] Provided:[%" PRId32 "]\n",
                        (int32_t)coll_id, 
                        (int32_t)opts.old_fsid);
        return -1;
    }

    if(opts.verbose)
    {
        printf("Retrieved key [%s] from db. New value is [%" PRId32 "]\n", 
               (char *)key.data, (int32_t)coll_id);
    }

    dbc_p->c_close(dbc_p);
    dbp->close(dbp, 0);
    return 0;
}

int process_args(int argc, char ** argv)
{
    int tmp_fsid = 0;
    int ret = 0, option_index = 0;
    static struct option long_opts[] =
    {
        {"help",0,0,0},
        {"verbose",0,0,0},
        {"oldfsid",1,0,0},
        {"newfsid",1,0,0},
        {"fsname",1,0,0},
        {"dbpath",1,0,0},
        {"fsconf",1,0,0},
        {"storage",1,0,0},
        {"view",0,0,0},
        {0,0,0,0}
    };

    memset(&opts, 0, sizeof(options_t));
    sprintf(opts.fs_name, "pvfs2-fs");

    while ((ret = getopt_long(argc, argv, "",
                              long_opts, &option_index)) != -1)
    {
	switch (option_index)
        {
            case 0: /* help */
                print_help(argv[0]);
                exit(0);

            case 1: /* verbose */
                opts.verbose = 1;
                break;

            case 2: /* oldfsid */
                tmp_fsid = atoi(optarg);
                opts.old_fsid = (int32_t) tmp_fsid;
                sprintf(opts.old_fsid_hex,"%08" PRIx32,opts.old_fsid);
                break;

            case 3: /* newfsid */
                tmp_fsid = atoi(optarg);
                opts.new_fsid = (int32_t) tmp_fsid;
                sprintf(opts.new_fsid_hex,"%08" PRIx32,opts.new_fsid);
                break;

            case 4: /* fsname */
                strncpy(opts.fs_name, optarg, PATH_MAX);
                break;

            case 5: /* dbpath */
                strncpy(opts.db_path, optarg, PATH_MAX);
                break;
            
            case 6: /* fsconf */
                strncpy(opts.fs_conf, optarg, PATH_MAX);
                break;

            case 7: /* storage */
                strncpy(opts.storage_path, optarg, PATH_MAX);
                break;

            case 8: /* view */
                opts.view_only = 1;
                break;

	    default:
                print_help(argv[0]);
		return(-1);
	}
        option_index = 0;
    }
    
    /* db_path must be set */
    if(!strncmp(opts.db_path,"",PATH_MAX))
    {
        fprintf(stderr,"Error: --dbpath option must be given.\n");
        print_help(argv[0]);
        return(-1);
    }

    /* fs_conf must be set */
    if(!strncmp(opts.fs_conf,"",PATH_MAX))
    {
        fprintf(stderr,"Error: --fsconf option must be given.\n");
        print_help(argv[0]);
        return(-1);
    }

    /* storage_path must be set */
    if(!strncmp(opts.storage_path,"",PATH_MAX))
    {
        fprintf(stderr,"Error: --storage option must be given.\n");
        print_help(argv[0]);
        return(-1);
    }

    return 0;
}

void print_help(char * progname)
{
    fprintf(stderr,"\nThis utility will update the fsid for a filesystem.\n");
    fprintf(stderr,"The following arguments are required:\n");
    fprintf(stderr,"--------------\n");
    fprintf(stderr,"  --dbpath=</path/to/collections.db>     "
            "The current file system ID.\n");
    fprintf(stderr,"  --fsconf=</path/to/pvfs2-fs.conf>     "
            "Fs config file for the the file system being modified.\n");
    fprintf(stderr,"  --storage=</path/to/pvfs2-storage-space>     "
            "Local storage space for the the file system being modified.\n");
    fprintf(stderr, "\n");
    fprintf(stderr,"The following arguments are optional:\n");
    fprintf(stderr,"--------------\n");
    fprintf(stderr,"  --oldfsid=<fs_id>     "
            "The current file system ID. Else looked up from fs conf.\n");
    fprintf(stderr,"  --newfsid=<fs_id>     "
            "The desired file system ID. Else generated as in genconfig.\n");
    fprintf(stderr,"  --verbose          "
            "Print verbose messages during execution.\n");
    fprintf(stderr,"  --help             "
            "Show this help listing.\n");
    fprintf(stderr, "\n");
    return;
}


/*
 * Local variables:
 *  c-indent-level: 4
 *  c-basic-offset: 4
 * End:
 *
 * vim: ts=8 sts=4 sw=4 expandtab
 */

